src/handler/mod.rs

```diff
@@ -1170,15 +1335,22 @@ impl Handler {
         }
     }
 
-    /// Sends a packet to the send handler to be encoded and sent.
+    /// Assembles and sends a [`Packet`].
     async fn send(&mut self, node_address: NodeAddress, packet: Packet) {
         let outbound_packet = socket::OutboundPacket {
             node_address,
             packet,
         };
-        if let Err(e) = self.socket.send.send(outbound_packet).await {
+        self.send_outbound(outbound_packet.into()).await;
+    }
+
+    /// Sends a packet to the send handler to be encoded and sent.
+    async fn send_outbound(&mut self, packet: Outbound) {
+        let dst = *packet.dst();
+        if let Err(e) = self.socket.send.send(packet).await {
             warn!("Failed to send outbound packet {}", e)
         }
+        self.nat_hole_puncher.track(dst);
```

Comment:
if the local node is behind a nat yes. holes must be kept open for all peers, see https://github.com/sigp/discv5/pull/176#discussion_r1441837283. a `LruTimeCache` is better to use here than `HashSetDelay` however, since frequently a sequence of partial nodes responses are sent, will fix that.